cpu/em_quad_level_frequency_transitions.py

```diff
@@ -0,0 +1,157 @@
+                    freq_set = self.cpu_freq_path('cpuinfo_cur_freq')
+                    if self.max_freq < freq_set:
+                        self.max_freq = freq_set
+                if chip not in self.max_freq_dict.keys():
+                    self.max_freq_dict[chip] = {}
+                if quad not in self.max_freq_dict[chip].keys():
+                    self.max_freq_dict[chip][quad] = self.max_freq
+                    self.log.info("Maximum frequency set:%s quad:"
+                                  "%s" % (self.max_freq, quad))
+                self.max_freq = 0
+        for chip in self.quad_dict:
+            for quad in self.quad_dict[chip]:
+                for cpu in self.quad_dict[chip][quad]:
+                    freq_get = self.perf_cmd(cpu)
+                    freq_max = self.max_freq_dict[chip][quad]
+                    diff = float(freq_max) - float(freq_get)
+                    if diff > threshold:
+                        self.cancel("Quad level max frequency %s is not set on"
+                                    "this cpu %s"
+                                    % (self.max_freq_dict[chip][quad], cpu))
+            self.log.info("Quad level max frequency %s is set on this quad"
+                          "%s" % (self.max_freq_dict[chip][quad], quad))
+
+    def quad_to_cpu_mapping(self):
+        """
+        Get the total quad and cpus list belonging to each quad.
+        """
+        self.nums = [x for x in range(0, self.cpus)]
+        for cpu in self.nums:
+            phy_id = process.system_output('cat /sys/devices/system/'
+                                           'cpu/cpu%s/physical_id' % cpu)
+            quad_id = int(phy_id) >> 4 & 0x7
+            chip_id = int(phy_id) >> 8 & 0x7F
+            if chip_id not in self.quad_dict.keys():
+                self.quad_dict[chip_id] = {}
+            if quad_id not in self.quad_dict[chip_id].keys():
+                self.quad_dict[chip_id][quad_id] = []
+            self.quad_dict[chip_id][quad_id].append(cpu)
+
+    def cpu_freq_path(self, file):
+        """
+        get cpu_freq values
+        :param: file: is filename for which data needs to be fetched
+        """
+        f_name = "/sys/devices/system/cpu/cpu%s/cpufreq/%s" \
+                 % (self.cpu_num, file)
+        return open(f_name, 'r').readline().strip('\n').strip(' ')
+
+    def get_random_freq(self):
+        """
+        Get random frequency from list
+        """
+        cmd = "scaling_available_frequencies"
+        return random.choice(self.cpu_freq_path(cmd).split(' '))
+
+    def perf_cmd(self, cpu):
+        """
+        Parse the output for perf cmd
+        :param: cpu: provide the output for each cpu
+        """
+        cmd = "perf stat -e cycles -e task-clock timeout 1 taskset -c \
+                %s yes>/dev/null" % cpu
+        output = process.run(cmd, shell=True, ignore_status=True)
+        self.log.info("output : %s" % output.stderr)
+        a = output.stderr
```

Comment:
please use meaningful  variable name as at least three char ?

Reply:
sure I will make this change